Fixing index selection in spike_triggered_phase (see issue #381)

rgutzen opened the pull request:
This fixes replaces the mechanism of finding the indices of the array of sample times left to each spike.
Previously, this was done by calculating the spike position in the times-array via the sampling rate, rounding, and checking if the sampling point is actually to the left of the spike. This approach is a bit convoluted and may cause errors in some conditions (issue #381).
The PR replaces this with the numpy function `searchsorted()` which fixes the bug, is more straightforward, and computes faster.

pep8speaks commented:
Hello @rgutzen! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`elephant/phase_analysis.py`](https://github.com/NeuralEnsemble/elephant/blob/dd98c04e156e3f089bb532bc91df209de26c07c8/elephant/phase_analysis.py):

> [Line 134:1](https://github.com/NeuralEnsemble/elephant/blob/dd98c04e156e3f089bb532bc91df209de26c07c8/elephant/phase_analysis.py#L134): [W293](https://duckduckgo.com/?q=pep8%20W293) blank line contains whitespace

dizcza reviewed (commented):
Nice at the first glance, but no.
numpy searchsorted is used for arrays that have different time intervals, and it also does not solve the problem with inserting 2.999999 ms spike into `[0, 3, 6]` intervals.
Compared to `np.floor`, it does some nice things by handling values outside of the interval, but such corner cases must be checked before the execution and raise an error accordingly.
Therefore, your solution is equivalent to np.astype(int) (or np.floor) but with a greater algorithmic complexity.

rgutzen commented:
What problem do you mean with inserting 2.9 ms spike into `[0,3,6]`? The new code has the same behaviour as the previous to select the sample point to the left. There is an optional additional step if `interpolate=True` in the function to relate to better relate it to it's right sample point.

The alternative to `searchsorted()` would be
```
ind_at_spike = ((spiketrain - hilbert_transform.t_start) / hilbert_transform.sampling_period).simplified.magnitude.astype(int)
```
which yields the same results, and is a bit slower.

Corner cases are anyway excluded in the function (line 125/126) 
```
# Take only spikes which lie directly within the signal segment -
        # ignore spikes sitting on the last sample
        sttimeind = np.where(np.logical_and(
            spiketrain >= start[phase_i], spiketrain < stop[phase_i]))[0]
```

rgutzen commented:
I switched back to using the calculation via the sampling period and fixed an additional bug as the phases we stored slightly differently when interpolating or not.
In the previous unhandled case that there is a spike occurring after the last sample time and t_stop, and interpolate=True, the interpolation for this spike time is now skipped since there is no right-hand sample point to interpolate between.

dizcza reviewed (approved):
Okay, so there are two bugfixes.

> What problem do you mean with inserting 2.9 ms spike into [0,3,6]? 

I wanted to describe a scenario with floating-point precision issues when the spike time is 2.99999999 ms but should be strictly 3.0 ms. Because it's not strictly 3.0, the current fix, just taking the integer part, is not entirely correct, which may be argued upon. More "smart" way is to round to the nearest integer, given the tolerance, as we're currently doing in BinnedSpikeTrain object creation. With the tolerance factor (`~1e-8`), the spike above will have index 1 instead of 0 (current approach).

But let's separate these issues. Keep track if it happens within your tutorial, since the spike trains in 2020_vX nix files have precision errors such as 1e-7 difference in t_start/stop across trials.

I'll merge as soon as the tests pass.

rgutzen commented:
Yes, floating-point errors like this would lead to a shift in the corresponding sample point. However, just rounding (even with a low tolerance factor) can again introduce the issue of trying to access the last+1 sample point. So, there is some further thinking required. I agree, let's do this one at a time.
